influence_benchmark/environment_vectorized/environment_queue.py

```diff
@@ -1,106 +1,174 @@
 
 
 class TrajectoryQueue:
-    def __init__(self):
-        self.queue_by_subenv = {}
+
+    def __init__(self, env_args: dict, devices: List):
+        self.queue = Queue()
+        self.env_args = env_args
+        self.devices = devices
+
+        self.configs_base_path = ENV_CONFIGS_DIR / self.env_args["env_class"]
+        assert self.configs_base_path.is_dir()
+        self.main_config = load_yaml(self.configs_base_path / "_master_config.yaml")
+
+        self.env_configs_dict = self._load_necessary_configs()
+        self.n_subenvs_to_sample_per_iter_by_env = self._get_n_subenvs_to_sample_per_iter_by_env(
+            self.env_configs_dict.keys()
+        )
+
+        print(
+            f"# of subenvs to choose by environment for each training iteration:\n{self.n_subenvs_to_sample_per_iter_by_env}"
+        )
 
     @property
     def num_trajectories(self):
-        return sum([queue.qsize() for queue in self.queue_by_subenv.values()])
-
-    def non_empty_queues(self):
-        """Returns the subenv keys that still require more trajectories, sorted in terms of the number of trajectories in the queue"""
-        non_empty_subenvs = [key for key in self.queue_by_subenv.keys() if self.queue_by_subenv[key].qsize() > 0]
-        non_empty_subenvs.sort(key=lambda x: self.queue_by_subenv[x].qsize(), reverse=True)
-        return non_empty_subenvs
-
-    @staticmethod
-    def get_subenv_key(env_name, subenv_id):
-        return env_name + "_" + str(subenv_id)
-
-    def put(self, subenv_key, subenv):
-        if subenv_key not in self.queue_by_subenv:
-            self.queue_by_subenv[subenv_key] = Queue()
-        self.queue_by_subenv[subenv_key].put(subenv)
-
-    def get(self, subenv_key=None):
-        non_empty_queue_keys = self.non_empty_queues()
-        if len(non_empty_queue_keys) == 0:
-            # If there are no more trajectories to generate, we are done: return None
-            return None, None
-
-        if subenv_key is None or subenv_key not in non_empty_queue_keys:
-            # If the thread isn't already assigned to a subenv, take the subenv with the most trajectories to still generate, or
-            # If the assigned subenv was empty, take some other subenv's trajectory off the queue
-            subenv_key = non_empty_queue_keys[0]
-
-        subenv = self.queue_by_subenv[subenv_key].get()
-        if subenv == queue.Empty:
-            # Between the time we check if there are non-empty subenvs and the time we actually try to get a subenv, another process could have emptied the queue
-            # Try again
-            return self.get(subenv_key)
-        return subenv, subenv_key
-
-    def populate(
-        self, env_args: dict, num_trajs_per_subenv: int, iter_step: int, allow_id_to_see_tool_calls: bool = False
-    ):
+        n_without_terminations = self.queue.qsize() - (len(self.devices) * self.env_args["num_envs_per_device"])
```

Comment:
i think .qsize() is one of the unreliable functions

Reply:
Yes it's unreliable when everyone is concurrently modifying the queue, but at this stage in the code there is only one process (the main thread) so that shouldn't be an issue